tools/cli/adminDBCommands.go

```diff
@@ -0,0 +1,264 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/urfave/cli"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/codec"
+	"github.com/uber/cadence/common/log/loggerimpl"
+	"github.com/uber/cadence/common/persistence"
+	cassp "github.com/uber/cadence/common/persistence/cassandra"
+)
+
+const (
+	scanWorkerCount            = 40
+	numberOfShards             = 16384
+	delayBetweenExecutionPages = 50 * time.Millisecond
+	executionPageSize          = 100
+	historyPageSize            = 50
+)
+
+type (
+	scanFiles struct {
+		failedToRunCheckFile  *os.File
+		startEventCorruptFile *os.File
+		progressFile          *os.File
+	}
+
+	// ExecutionScanEntity is the execution entity which gets written to output file from scan
+	ExecutionScanEntity struct {
+		ShardID      int
+		DomainID     string
+		WorkflowID   string
+		RunID        string
+		NextEventID  int64
+		TreeID       string
+		BranchID     string
+		CloseStatus  int
+		ScanMetadata ExecutionScanEntityMetadata
+	}
+
+	// ExecutionScanEntityMetadata is the metadata from scanning the execution that gets written to output file
+	ExecutionScanEntityMetadata struct {
+		Message  string
+		ErrorMsg string
+	}
+
+	// ScanShardReport contains progress information about the scan, this struct is used
+	// both to report global progress and per shard reports.
+	ScanShardReport struct {
+		ShardID                     int
+		NumberOfExecutions          int
+		NumberOfCorruptedExecutions int
+		NumberOfFailedChecks        int
+		FailedToListExecutions      bool
+	}
+)
+
+// AdminDBScan is used to scan over all executions in database and detect corruptions
+func AdminDBScan(c *cli.Context) {
+	scanFiles, deferFn := createScanFiles()
+	session := connectToCassandra(c)
+	defer func() {
+		deferFn()
+		session.Close()
+	}()
+
+	shardReports := make(chan *ScanShardReport)
+	for i := 0; i < scanWorkerCount; i++ {
+		go func(workerIdx int) {
+			for shardID := 0; shardID < numberOfShards; shardID++ {
+				if shardID%scanWorkerCount == workerIdx {
+					shardReports <- scanShard(session, shardID, scanFiles)
+				}
+			}
+		}(i)
```

Comment:
Is it possible to stop/terminate the scan after it's started?

Reply:
Just kill -9 the process